src/Microsoft.AspNetCore.Routing/Tree/TreeRouteBuilder.cs

```diff
@@ -25,17 +25,46 @@ public class TreeRouteBuilder
         private readonly IInlineConstraintResolver _constraintResolver;
 
         /// <summary>
-        /// Initializes a new instance of <see cref="TreeRouteBuilder"/>.
+        /// <para>
+        /// This constructor is obsolete and will be removed in a future version. The recommended
+        /// alternative is the overload that does not take a UrlEncoder.
+        /// </para>
+        /// <para>Initializes a new instance of <see cref="TreeRouteBuilder"/>.</para>
         /// </summary>
         /// <param name="loggerFactory">The <see cref="ILoggerFactory"/>.</param>
         /// <param name="urlEncoder">The <see cref="UrlEncoder"/>.</param>
         /// <param name="objectPool">The <see cref="ObjectPool{UrlBuildingContext}"/>.</param>
         /// <param name="constraintResolver">The <see cref="IInlineConstraintResolver"/>.</param>
+        [Obsolete("This constructor is obsolete and will be removed in a future version. The recommended " +
+            "alternative is the overload that does not take a UrlEncoder.")]
         public TreeRouteBuilder(
             ILoggerFactory loggerFactory,
             UrlEncoder urlEncoder,
             ObjectPool<UriBuildingContext> objectPool,
             IInlineConstraintResolver constraintResolver)
+            : this(loggerFactory, objectPool, constraintResolver, urlEncoder)
```

Comment:
All comments made in `TemplateBinder` apply to this class as well. Don't need more than one new constructor.